error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MemberOrdering.java

```diff
@@ -0,0 +1,187 @@
+package tech.picnic.errorprone.bugpatterns;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.errorprone.BugPattern.LinkType.CUSTOM;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static com.google.errorprone.BugPattern.StandardTags.STYLE;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static tech.picnic.errorprone.bugpatterns.util.Documentation.BUG_PATTERNS_BASE_URL;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.google.errorprone.util.ErrorProneToken;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.parser.Tokens;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+import javax.lang.model.element.Modifier;
+
+/** A {@link BugChecker} that flags classes with non-standard member ordering. */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Members should be ordered in a standard way",
+    explanation =
+        "Members should be ordered in a standard way, which is: "
+            + "static member variables, non-static member variables, constructors and methods.",
+    link = BUG_PATTERNS_BASE_URL + "MemberOrdering",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = STYLE)
+public final class MemberOrdering extends BugChecker implements BugChecker.ClassTreeMatcher {
+  private static final long serialVersionUID = 1L;
+
+  /** A comparator that sorts variable and method (incl. constructors) in a standard order. */
+  private static final Comparator<Tree> MEMBER_SORTING =
+      comparing(
+          (Tree tree) -> {
```

Comment:
Saw this being renamed to `tree` from `memberTree`, but in other placed we renamed `memberTree` to `classMemberTree`.

Personally I don't mind either. I originally went with member tree because afaik, only classes have members in Java, specifying `class` would be redundant. One thing to note, if we go with `classMember*`, we should rename this BugChecker to `ClassMemberOrdering`

Reply:
> Saw this being renamed to tree from memberTree while in other places we renamed memberTree to classMemberTree. Do you think we can use the same strategy everywhere?

I think this is purely based on the type we're referring to, e.g. `Tree` -> `tree`.
And since we renamed `MemberTreeWithComments` to `ClassMemberTreeWithComments` the variable names should reflect that.

> I originally went with member tree because afaik, only classes have members in Java, specifying class would be redundant.

`ClassMemberTreeWithComments` is a name @rickie preferred. I think it does add some clarity. That being said, [JEP 445](https://openjdk.org/jeps/445) introduces unnamed classes which could be a concern for the `MemberTreeWithComments` name :joy: